apps/system/js/sync_state_machine.js

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+
+       +---------------Success ---------------+
+       |                                      |
+   +---+----+                                 |
+   |Enabling+---Error---+                     |
+   +---+----+           |                     |
+       ^              +-v---+                 |
+       |   +-Disable--+Error+-------------+   |
+       |   |          +--+--+             |   |
+     Enable|             ^             Enable |
+       |   |             |                |   |
+       |   |             |                |   |
+  +----+---v-+           |            +---v---v+
+  | Disabled |         Error +------->+Enabled +---+
+  +----^-----+           |   |        +---+----+   |
+       |                 |   |            |        |
+       |                 | Success       Sync      |
+       |                 |   |            |        |
+       |             +---+---+            |        |
+       |             |Syncing<------------+        |
+       |             +-------+                     |
+       |                                           |
+       +--------------Disable  --------------------+
+
+*/
+
+'use strict';
+
+(function(exports) {
+
+  var SyncStateMachine = {
+    current: 'disabled'
+  };
+
+  /**
+   * Object representing the possible events and the associated states.
+   * {
+   *   eventName: {
+   *     from: to,
+   *     from: to
+   *     ...
+   *   },
+   *   ...
+   * }
+   */
+  const events = {
+    enable: {
+      'disabled': 'enabling',
+      'error': 'enabled'
```

Comment:
Why:
    disabled ---enable---> enabling
    error ---enable---> enabled

and not:
    disabled ---enable---> enabling
    error ---enable---> enabling

?